impl/blockservice/fetcher.go

```diff
@@ -97,10 +97,14 @@ func (f *fetcherSession) NodeMatching(ctx context.Context, node ipld.Node, match
 }
 
 func (f *fetcherSession) BlockMatchingOfType(ctx context.Context, root ipld.Link, match ipld.Node,
-	ptype ipld.NodePrototype, cb fetcher.FetchCallback) error {
+	_ ipld.NodePrototype, cb fetcher.FetchCallback) error {
 
 	// retrieve first node
-	node, err := f.BlockOfType(ctx, root, ptype)
+	prototype, err := f.PrototypeFromLink(root)
```

Comment:
Special casing the type of the first node over the others seems harmful/bug prone. We can just use the linksystem to figure out what it should be.

It seems like we could probably also just change the name and signature of this function in the Fetcher interface to drop this. I'm not sure if it would cause much propagation damage and should wait for a follow-up set of PRs though. If we're not going to do this let's at least add a comment to the interface saying that the prototype should be ignored.

Reply:
I'm not sure what the issue is here?  And I'm not sure LinkSystem is that smart?

One typically wants a NodePrototype parameter to either A) say what kind of memory layout to use, or B) say what kind of types to use, if it's a schema node (and this, then, can go on to hint where ADLs should be later, etc, though we do not use this anywhere in IPFS yet, and it may be some time before we do).

LinkSystem has a NodeReifier callback, yes, but it's not meant to do half of those jobs.

(All of this should be prefixed with "I think".  I don't understand big picture what this function is supposed to be doing.)

Reply:
... Looking yet again, I set the strength of my previous comment to zero.  Because I agree with [Will's comment](https://github.com/ipfs/go-fetcher/pull/24#issuecomment-899009262) that `PrototypeFromLink` doesn't seem to belong here at all, I just can't make sense of this, so if you're telling me this fixed something, I can't protest.